[SPARK-4563][core] Allow driver to advertise a different network address.

vanzin opened the pull request:
The goal of this feature is to allow the Spark driver to run in an
isolated environment, such as a docker container, and be able to use
the host's port forwarding mechanism to be able to accept connections
from the outside world.

The change is restricted to the driver: there is no support for achieving
the same thing on executors (or the YARN AM for that matter). Those still
need full access to the outside world so that, for example, connections
can be made to an executor's block manager.

The core of the change is simple: add a new configuration that tells what's
the address the driver should bind to, which can be different than the address
it advertises to executors (spark.driver.host). Everything else is plumbing
the new configuration where it's needed.

To use the feature, the host starting the container needs to set up the
driver's port range to fall into a range that is being forwarded; this
required the block manager port to need a special configuration just for
the driver, which falls back to the existing spark.blockManager.port when
not set. This way, users can modify the driver settings without affecting
the executors; it would theoretically be nice to also have different
retry counts for driver and executors, but given that docker (at least)
allows forwarding port ranges, we can probably live without that for now.

Because of the nature of the feature it's kinda hard to add unit tests;
I just added a simple one to make sure the configuration works.

This was tested with a docker image running spark-shell with the following
command:

 docker blah blah blah \
   -p 38000-38100:38000-38100 \
   [image] \
   spark-shell \
     --num-executors 3 \
     --conf spark.shuffle.service.enabled=false \
     --conf spark.dynamicAllocation.enabled=false \
     --conf spark.driver.host=[host's address] \
     --conf spark.driver.port=38000 \
     --conf spark.driver.blockManager.port=38020 \
     --conf spark.ui.port=38040

Running on YARN; verified the driver works, executors start up and listen
on ephemeral ports (instead of using the driver's config), and that caching
and shuffling (without the shuffle service) works. Clicked through the UI
to make sure all pages (including executor thread dumps) worked. Also tested
apps without docker, and ran unit tests.

Commit:
[SPARK-4563][core] Allow driver to advertise a different network address.

The goal of this feature is to allow the Spark driver to run in an
isolated environment, such as a docker container, and be able to use
the host's port forwarding mechanism to be able to accept connections
from the outside world.

The change is restricted to the driver: there is no support for achieving
the same thing on executors (or the YARN AM for that matter). Those still
need full access to the outside world so that, for example, connections
can be made to an executor's block manager.

The core of the change is simple: add a new configuration that tells what's
the address the driver should bind to, which can be different than the address
it advertises to executors (spark.driver.host). Everything else is plumbing
the new configuration where it's needed.

To use the feature, the host starting the container needs to set up the
driver's port range to fall into a range that is being forwarded; this
required the block manager port to need a special configuration just for
the driver, which falls back to the existing spark.blockManager.port when
not set. This way, users can modify the driver settings without affecting
the executors; it would theoretically be nice to also have different
retry counts for driver and executors, but given that docker (at least)
allows forwarding port ranges, we can probably live without that for now.

Because of the nature of the feature it's kinda hard to add unit tests;
I just added a simple one to make sure the configuration works.

This was tested with a docker image running spark-shell with the following
command:

 docker blah blah blah \
   -p 38000-38100:38000-38100 \
   [image] \
   spark-shell \
     --num-executors 3 \
     --conf spark.shuffle.service.enabled=false \
     --conf spark.dynamicAllocation.enabled=false \
     --conf spark.driver.host=[host's address] \
     --conf spark.driver.port=38000 \
     --conf spark.driver.blockManager.port=38020 \
     --conf spark.ui.port=38040

Running on YARN; verified the driver works, executors start up and listen
on ephemeral ports (instead of using the driver's config), and that caching
and shuffling (without the shuffle service) works. Clicked through the UI
to make sure all pages (including executor thread dumps) worked. Also tested
apps without docker, and ran unit tests.

SparkQA commented:
**[Test build #65509 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65509/consoleFull)** for PR 15120 at commit [`40f23a4`](https://github.com/apache/spark/commit/40f23a48a8eca3a7d20a193a7998413090ea695c).

SparkQA commented:
**[Test build #65509 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65509/consoleFull)** for PR 15120 at commit [`40f23a4`](https://github.com/apache/spark/commit/40f23a48a8eca3a7d20a193a7998413090ea695c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65509/
Test FAILed.

Commit:
Fix CheckpointSuite.

SparkQA commented:
**[Test build #65516 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65516/consoleFull)** for PR 15120 at commit [`0c1fe29`](https://github.com/apache/spark/commit/0c1fe29e4fea8782f16e3f6d2524f5b781e63e34).

SparkQA commented:
**[Test build #65516 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65516/consoleFull)** for PR 15120 at commit [`0c1fe29`](https://github.com/apache/spark/commit/0c1fe29e4fea8782f16e3f6d2524f5b781e63e34).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65516/
Test PASSed.

vanzin commented:
/cc @zsxwing

zsxwing commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -143,4 +144,23 @@ package object config {
+  private[spark] val DRIVER_BIND_ADDRESS = ConfigBuilder("spark.driver.bindAddress")
+    .doc("Address where to bind network listen sockets on the driver.")
+    .stringConf
+    .createWithDefault(Utils.localHostName())
```
This is a broken change. If a user uses `spark.driver.host` to specify the bind address, it won't work now. Right?

vanzin replied:
It should, as long as he doesn't set "spark.driver.bindAddress" - which is a new setting that is being added to override that behavior.

zsxwing replied:
Maybe I miss something. `val bindAddress = conf.get(DRIVER_BIND_ADDRESS)` won't use `spark.driver.host`. Right?

vanzin replied:
Ah I see what you mean. I might have inverted the config resolution order... let me take a look.

Commit:
Bind address should default to spark.driver.host.

And not the other way around, for backwards compatibility. Includes
a small change to the resolution of fallback conf entry values.

SparkQA commented:
**[Test build #65692 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65692/consoleFull)** for PR 15120 at commit [`5137131`](https://github.com/apache/spark/commit/513713144ef36351663ec03290ec09dc99f5c01c).

SparkQA commented:
**[Test build #65692 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65692/consoleFull)** for PR 15120 at commit [`5137131`](https://github.com/apache/spark/commit/513713144ef36351663ec03290ec09dc99f5c01c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65692/
Test FAILed.

zsxwing commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2079,9 +2079,9 @@ private[spark] object Utils extends Logging {
-              s"service$serviceString (for example spark.ui.port for SparkUI) to an available " +
-              "port or increasing spark.port.maxRetries."
+              s"$maxRetries retries (starting from $startPort)! Consider explicitly setting " +
+              s"the appropriate port for the service$serviceString (for example spark.ui.port " +
```
nit: since you are touching this, could you add a space between `service` and `$serviceString`?

vanzin replied:
The space is actually in `$serviceString` because the service name is allowed to be empty. (Historical reasons? Don't ask me.)

zsxwing replied:
Got it.

zsxwing commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -383,8 +383,9 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
-    _conf.setIfMissing("spark.driver.host", Utils.localHostName())
+    // Set Spark driver host and port system properties. This explicitly sets the configuration
+    // instead of relying on the default value of the config constant.
+    _conf.set(DRIVER_HOST_ADDRESS, conf.get(DRIVER_HOST_ADDRESS))
```
nit: could you use `_conf` instead? I thought `_conf` and `conf` were different at first glance. But actually they are same.

zsxwing commented on core/src/main/scala/org/apache/spark/internal/config/ConfigProvider.scala:
```diff
@@ -66,7 +66,7 @@ private[spark] class SparkConfigProvider(conf: JMap[String, String]) extends Con
       case e: ConfigEntryWithDefault[_] => Option(e.defaultValueString)
       case e: ConfigEntryWithDefaultString[_] => Option(e.defaultValueString)
-      case e: FallbackConfigEntry[_] => defaultValueString(e.fallback.key)
+      case e: FallbackConfigEntry[_] => get(e.fallback.key)
```
do we need to backport this fix to 2.0.1?

vanzin replied:
No, this code is only in master.

sumitvashistha replied:
We are facing this issue with Spark 1.6 . Are we going to backport this?

vanzin replied:
What issue? The code you're commenting on does not exist in 1.6. If you're having issues, please ask questions on the mailing lists or use the bug tracker.

zsxwing reviewed (changes requested):
LGTM. Just some nits

Commit:
Use '_conf' consistently.

SparkQA commented:
**[Test build #65726 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65726/consoleFull)** for PR 15120 at commit [`6dd0716`](https://github.com/apache/spark/commit/6dd071681a4bb967d4a0f86be29c8781efb40fbe).

SparkQA commented:
**[Test build #65726 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65726/consoleFull)** for PR 15120 at commit [`6dd0716`](https://github.com/apache/spark/commit/6dd071681a4bb967d4a0f86be29c8781efb40fbe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65726/
Test PASSed.

zsxwing commented:
Merging to master. Thanks!

eduardohl commented:
Hi, I'm currently in need of using this change because we're running inside containers that need port mapping. We're not using master (currently using 2.0.2). Do you know if there are any alternatives while the commit isn't released?

vanzin commented:
> Do you know if there are any alternatives while the commit isn't released?

Not really.

cherryii commented:
I am using spark 2.1.0, my docker container is a spark driver that only runs my job, my container's entrypoint is a script that runs spark submit. I use the client deploy-mode, set the conf values defined above and get the error below when in bridge networking, does anyone know what I'm doing wrong or why bridge networking isn't working with the conf values defined above?

java.net.BindException: Cannot assign requested address: Service 'sparkDriver' failed after 16 retries! Consider explicitly setting the appropriate port for the service 'sparkDriver' (for example spark.ui.port for SparkUI) to an available port or increasing spark.port.maxRetries.
at sun.nio.ch.Net.bind0(Native Method)
at sun.nio.ch.Net.bind(Net.java:433)
at sun.nio.ch.Net.bind(Net.java:425)
at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
at io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:125)
at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:485)
at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1089)
at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:430)
at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:415)
at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:903)
at io.netty.channel.AbstractChannel.bind(AbstractChannel.java:198)
at io.netty.bootstrap.AbstractBootstrap$2.run(AbstractBootstrap.java:348)
at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:357)
at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:357)
at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
at java.lang.Thread.run(Thread.java:745)

sangramga commented:
@SparkQA Is this feature available for only YARN cluster manager? Can Spark standalone manager can also use this?
